zerver/views/auth.py

```diff
@@ -216,8 +216,19 @@ def remote_user_sso(request: HttpRequest,
     # Since RemoteUserBackend will return None if Realm is None, we
     # don't need to check whether `get_realm` returned None.
     user_profile = authenticate(remote_user=remote_user, realm=realm)
+
+    next = request.GET.get('next', '')
+    redirect_to = next
+    # We decide whether the next link is a narrow or not based on the fact that
+    # either the path will be relative and hence begin with a '/' or be a narrow
+    # in which case it should have had a '#' in the start but we removed that
+    # using Js so that the narrow path could make it to the backend.
+    if not redirect_to.startswith('/') and redirect_to != '':
+        redirect_to = '/#' + redirect_to
```

Comment:
This feels a bit messy; couldn't the path be `/#narrow`, for example?  Again, I think using a `urllib.parse` command and then actually inspecting the `.path` part would be a cleaner implementation.

Reply:
So actually redirect_to for the SSO path can't really contain a `#` ever since browser never sends the hash to backend. For the google and social code path's I introduced hidden input variables which were used to get the hash to the backend. 
For SSO I couldn't really do that because it is based on visiting a `<a>` hyperlink. Since I needed to get the hash to the backend what I did on the front end was that append the hash after removing the `#` to the href link.

Href link looks like this `href="/accounts/login/sso/?next={{ next }}"` and appending the hash resulted in the next variable carry the param to backend. 

In the above code I separate the hash and other next url's which could only be uploaded files (I couldn't think of any other think opening up from our app using the next code path.). Since the uploads will have a `/` in there next path and for the narrow's there won't be a `/`.